From 382acb891973d36787a66509a81c97d4e5605391 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:18:49 -0400
Subject: [fs] nfsd: reduce some spinlocking in put_client_renew

Message-id: <1410898849-22126-46-git-send-email-bfields@redhat.com>
Patchwork-id: 95716
O-Subject: [Patch RHEL7 045/165] nfsd: reduce some spinlocking in put_client_renew
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Jeff Layton <jlayton@primarydata.com>

No need to take the lock unless the count goes to 0.

Signed-off-by: Jeff Layton <jlayton@primarydata.com>
Reviewed-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: d6c249b4d4cfef894cbda224a7a063d17aacb60a
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 63c1420..3704789 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -193,8 +193,10 @@ static void put_client_renew(struct nfs4_client *clp)
 {
  struct nfsd_net *nn = net_generic(clp->net, nfsd_net_id);
 
- spin_lock(&nn->client_lock);
- put_client_renew_locked(clp);
+ if (!atomic_dec_and_lock(&clp->cl_refcount, &nn->client_lock))
+  return;
+ if (!is_client_expired(clp))
+  renew_client_locked(clp);
  spin_unlock(&nn->client_lock);
 }
 
-- 
1.7.1